tools/app.hpp

```diff
@@ -0,0 +1,184 @@
+#pragma once
+
+#include <optional>
+#include <string>
+#include <thread>
+
+#include <CLI/CLI.hpp>
+#include <range/v3/view/getlines.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include "io.hpp"
+#include "query/queries.hpp"
+
+namespace pisa {
+
+namespace arg {
+
+    enum class Encoding : bool { Legacy, New };
+
+    template <Encoding encoding = Encoding::New>
+    struct Index {
+        explicit Index(CLI::App *app)
+        {
+            app->add_option("-i,--index", m_basename, "Inverted index basename")->required();
+            if constexpr (encoding == Encoding::New) {
+                app->add_option("-e,--encoding", m_encoding, "Index encoding")->required();
+            } else {
+                app->add_option("-e,-t,--encoding,--type",
+                                m_encoding,
+                                "Index encoding\nDEPRECATED: -t/--type; USE: -e/--encoding")
+                    ->required();
+            }
+            app->add_option("-w,--wand", m_wand_data_path, "WAND data filename");
+            app->add_flag("--compressed-wand", m_wand_compressed, "Compressed WAND data file");
+        }
+
+        [[nodiscard]] auto index_basename() const -> std::string const & { return m_basename; }
+        [[nodiscard]] auto index_encoding() const -> std::string const & { return m_encoding; }
+        [[nodiscard]] auto wand_data_path() const -> std::optional<std::string> const &
+        {
+            return m_wand_data_path;
+        }
+        [[nodiscard]] auto is_wand_compressed() const -> bool { return m_wand_compressed; }
+
+       private:
+        std::string m_basename;
+        std::string m_encoding;
+        std::optional<std::string> m_wand_data_path;
+        bool m_wand_compressed = false;
+    };
+
+    enum class QueryMode : bool { Ranked, Unranked };
+
+    template <QueryMode Mode = QueryMode::Ranked, int DefaultK = 1'000>
+    struct Query {
+        explicit Query(CLI::App *app)
+        {
+            app->add_option("-q,--queries", m_query_file, "Path to file with queries", false);
+            app->add_option(
+                "--stopwords", m_stop_words, "List of blacklisted stop words to filter out");
```

Comment:
Depends on `queries` and maybe `terms`?
You pass stopwords, but than you dont pass any `stemmer` or `terms` file. What happens?